proposals/p2188.md

```diff
@@ -0,0 +1,517 @@
+
+## Proposal
+
+Patterns in Carbon are a generalization of the expression grammar. Compared to
+expressions, patterns add:
+
+-   Bindings, of the form `name: type`, which give a name for the scrutinee.
+-   `var` _pattern_, which requests mutable storage be provided for the
+    scrutinee.
+-   Additional syntax to make matching against structs more convenient.
+
+## Details
+
+### Expressions versus proper patterns
+
+Expressions are patterns, as described below. A pattern that is not an
+expression, because it contains pattern-specific syntax such as a binding, is a
+_proper pattern_. Many expression forms, such as arbitrary function calls, are
+not permitted as proper patterns, so cannot contain bindings.
+
+-   _pattern_ ::= _proper-pattern_
+
+```
+fn F(n: i32) -> i32 { return n; }
+
+match (F(42)) {
+  // ❌ Error: binding can't appear in a function call.
+  case (F(n: i32)) => {}
+}
+```
+
+### Expression patterns
+
+An expression is a pattern.
+
+-   _pattern_ ::= _expression_
+
+The pattern is compared with the expression using the `==` operator: _pattern_
+`==` _scrutinee_.
+
+```
+fn F(n: i32) {
+  match (n) {
+    // ✅ Results in an `n == 5` comparison.
+    // OK despite `n` and `5` having different types.
+    case 5 => {}
+  }
+}
+```
+
+Any `==` operations performed by a pattern match occur in lexical order, but for
+repeated matches against the same `_pattern_`, later comparisons may be skipped
+by reusing the result from an earlier comparison:
+
+```
+class ChattyIntMatcher {
+  external impl as EqWith(i32) {
+    fn Eq[me: ChattyIntMatcher](other: i32) {
+      Print("Matching {0}", other);
+      return other == 1;
+    }
+  }
+}
+
+fn F() {
+  // Prints `Matching 1` then `Matching 2`,
+  // may or may not then print `Matching 1` again.
+  match ((1, 2)) {
+    case ({} as ChattyIntMatcher, 0) => {}
+    case (1, {} as ChattyIntMatcher) => {}
+    case ({} as ChattyIntMatcher, 2) => {}
+  }
+}
+```
+
+### Bindings
+
+A name binding is a pattern.
+
+-   _binding-pattern_ ::= `unused`? _identifier_ `:` _expression_
+-   _binding-pattern_ ::= `unused`? _identifier_ `:` _proper-pattern_
+-   _proper-pattern_ ::= _binding-pattern_
+
+In the first form, the type of the _identifier_ is specified by the
+_expression_. In the second form, the _proper-pattern_ is matched against the
+type of the scrutinee, and the result is the type of the _identifier_. In either
+case, the scrutinee is implicitly converted to the type of the _identifier_.
+
+```
+fn F(p: i32*) {
+  match (p) {
+    // ✅ Matches `(T:! Type)` against `i32*`.
+    // `T` is bound to `i32`, `m` is bound to `p`, and the type of `m` is `i32*`.
+    case m: (T:! Type)* => {}
+  }
+}
+```
+
+```
+fn G() -> i32 {
+  match (5) (
+    // ✅ `5` is implicitly converted to `i32`.
+    // Returns `5 as i32`.
+    case n: i32 => { return n; }
+  }
+}
+```
+
+The `unused` keyword indicates that the binding is intended to not be used.
```

Comment:
It might be a good idea to explicitly say that this is intended to restate #2022's decision, not supersede it, just in case we miss some subtle disagreement between them.